File: lib/services/notification_helper.dart
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:flutter/material.dart';
import 'package:permission_handler/permission_handler.dart';
import 'dart:io';

class NotificationHelper {
  static final FlutterLocalNotificationsPlugin _notifications = FlutterLocalNotificationsPlugin();

  static const String receiveChannelId = 'receive_channel';
  static const String sendChannelId = 'send_channel';
  static const String welcomeChannelId = 'welcome_channel';
  static const String priceAlertChannelId = 'price_alert_channel';
  static const String loginChannelId = 'login_channel_id';
  static const int notificationId = 1;

  /// مقداردهی اولیه کانال‌ها
  static Future<void> initialize() async {
    const AndroidInitializationSettings androidInit = AndroidInitializationSettings('notifsmall');
    const DarwinInitializationSettings iosInit = DarwinInitializationSettings();
    const InitializationSettings initSettings = InitializationSettings(android: androidInit, iOS: iosInit);
    await _notifications.initialize(initSettings);
    await _createNotificationChannels();
  }

  /// ساخت کانال‌های نوتیفیکیشن
  static Future<void> _createNotificationChannels() async {
    if (Platform.isAndroid) {
      final android = _notifications.resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>();
      if (android != null) {
        await android.createNotificationChannel(const AndroidNotificationChannel(
          receiveChannelId,
          'Receive Notifications',
          description: 'Channel for receive notifications',
          importance: Importance.high,
          sound: RawResourceAndroidNotificationSound('receive_sound'),
        ));
        await android.createNotificationChannel(const AndroidNotificationChannel(
          sendChannelId,
          'Send Notifications',
          description: 'Channel for send notifications',
          importance: Importance.high,
          sound: RawResourceAndroidNotificationSound('send_sound'),
        ));
        await android.createNotificationChannel(const AndroidNotificationChannel(
          welcomeChannelId,
          'Welcome Notifications',
          description: 'Channel for welcome notifications',
          importance: Importance.defaultImportance,
          sound: RawResourceAndroidNotificationSound('welcome_sound'),
        ));
        await android.createNotificationChannel(const AndroidNotificationChannel(
          priceAlertChannelId,
          'Price Alert Notifications',
          description: 'Channel for price alert notifications',
          importance: Importance.high,
          sound: RawResourceAndroidNotificationSound('price_alert_sound'),
        ));
        await android.createNotificationChannel(const AndroidNotificationChannel(
          loginChannelId,
          'Login Notifications',
          description: 'Channel for login notifications',
          importance: Importance.defaultImportance,
        ));
      }
    }
  }

  /// درخواست مجوز نوتیفیکیشن (Android 13+ و iOS)
  static Future<void> requestNotificationPermission(BuildContext context) async {
    if (Platform.isAndroid) {
      final status = await Permission.notification.status;
      if (!status.isGranted) {
        await Permission.notification.request();
      }
    } else if (Platform.isIOS) {
      // iOS: مجوز هنگام initialize درخواست می‌شود
    }
  }

  /// ارسال نوتیفیکیشن ساده
  static Future<void> showNotification({
    required String channelId,
    required String title,
    required String body,
    String? payload,
    String? largeIconAsset,
  }) async {
    final androidDetails = AndroidNotificationDetails(
      channelId,
      channelId,
      channelDescription: 'Channel for $channelId',
      importance: Importance.high,
      priority: Priority.high,
      largeIcon: largeIconAsset != null ? DrawableResourceAndroidBitmap(largeIconAsset) : null,
    );
    final iosDetails = const DarwinNotificationDetails();
    final details = NotificationDetails(android: androidDetails, iOS: iosDetails);
    await _notifications.show(
      DateTime.now().millisecondsSinceEpoch % 100000,
      title,
      body,
      details,
      payload: payload,
    );
  }

  /// ارسال نوتیفیکیشن خوش‌آمد
  static Future<void> showWelcomeNotification() async {
    await showNotification(
      channelId: welcomeChannelId,
      title: 'Welcome',
      body: 'Welcome to ADL Wallet',
      largeIconAsset: 'logo',
    );
  }

  /// ارسال نوتیفیکیشن دریافت وجه
  static Future<void> showReceiveNotification(double amount, String currency) async {
    await showNotification(
      channelId: receiveChannelId,
      title: 'دریافت وجه',
      body: 'شما $amount $currency دریافت کردید!',
      largeIconAsset: 'logo',
    );
  }

  /// ارسال نوتیفیکیشن ارسال وجه
  static Future<void> showSendNotification(double amount, String currency) async {
    await showNotification(
      channelId: sendChannelId,
      title: 'ارسال وجه',
      body: 'شما $amount $currency ارسال کردید!',
      largeIconAsset: 'logo',
    );
  }

  /// ارسال نوتیفیکیشن هشدار قیمت
  static Future<void> showPriceAlertNotification(double currentPrice) async {
    await showNotification(
      channelId: priceAlertChannelId,
      title: 'هشدار قیمت!',
      body: 'قیمت بیت‌کوین به $currentPrice دلار رسید!',
      largeIconAsset: 'logo',
    );
  }

  /// حذف همه نوتیفیکیشن‌ها
  static Future<void> cancelAllNotifications() async {
    await _notifications.cancelAll();
  }

  /// حذف کانال‌های نوتیفیکیشن (فقط اندروید)
  static Future<void> deleteNotificationChannels() async {
    if (Platform.isAndroid) {
      final android = _notifications.resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>();
      if (android != null) {
        await android.deleteNotificationChannel(receiveChannelId);
        await android.deleteNotificationChannel(sendChannelId);
        await android.deleteNotificationChannel(welcomeChannelId);
        await android.deleteNotificationChannel(priceAlertChannelId);
        await android.deleteNotificationChannel(loginChannelId);
      }
    }
  }

  /// Initialize notification settings (stub method for compatibility)
  static Future<void> initializeNotificationSettings() async {
    try {
      print('📱 Initializing notification settings...');
      
      // Request notification permissions on iOS and Android 13+
      if (Platform.isIOS || (Platform.isAndroid)) {
        final permission = await Permission.notification.request();
        if (permission.isGranted) {
          print('✅ Notification permission granted');
        } else {
          print('❌ Notification permission denied');
        }
      }
      
      print('✅ Notification settings initialized');
    } catch (e) {
      print('❌ Error initializing notification settings: $e');
    }
  }
} 

File: 1/NotificationHelper.kt
package com.laxce.adl.utility

import android.app.*
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.media.AudioAttributes
import android.net.Uri
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import android.Manifest
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import com.laxce.adl.MainActivity
import com.laxce.adl.R
import android.app.NotificationChannel
import android.app.NotificationManager

class NotificationHelper(private val context: Context) {

    fun cancelAllNotifications() {
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.cancelAll()
    }

    fun deleteNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.deleteNotificationChannel(WELCOME_CHANNEL_ID)
            notificationManager.deleteNotificationChannel(RECEIVE_CHANNEL_ID)
            notificationManager.deleteNotificationChannel(SEND_CHANNEL_ID)
            notificationManager.deleteNotificationChannel(PRICE_ALERT_CHANNEL_ID)
        }
    }


    companion object {
        const val RECEIVE_CHANNEL_ID = "receive_channel"
        const val SEND_CHANNEL_ID = "send_channel"
        const val WELCOME_CHANNEL_ID = "welcome_channel"
        const val PRICE_ALERT_CHANNEL_ID = "price_alert_channel"
        const val CHANNEL_ID = "login_channel_id"
        const val NOTIFICATION_ID = 1
    }


    fun createNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createChannel(RECEIVE_CHANNEL_ID, "Receive Notifications", "receive_sound")
            createChannel(SEND_CHANNEL_ID, "Send Notifications", "send_sound")
            createChannel(WELCOME_CHANNEL_ID, "Welcome Notifications", "welcome_sound")
            createChannel(PRICE_ALERT_CHANNEL_ID, "Price Alert Notifications", "price_alert_sound")
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Login Notifications",
                NotificationManager.IMPORTANCE_DEFAULT
            ).apply {
                description = "Notifications for user login"
            }

            val notificationManager = context.getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun createChannel(channelId: String, channelName: String, soundFileName: String) {
        val soundUri = Uri.parse("android.resource://${context.packageName}/raw/$soundFileName")

        val attributes = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_NOTIFICATION)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .build()

        val channel = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_HIGH).apply {
            description = "Channel for $channelName"
            setSound(soundUri, attributes)
        }

        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.createNotificationChannel(channel)
    }


    private fun getSoundUri(soundFileName: String): Uri? {
        val resourceId = context.resources.getIdentifier(soundFileName, "raw", context.packageName)
        return if (resourceId != 0) {
            Uri.parse("android.resource://${context.packageName}/raw/$soundFileName")
        } else {
            null
        }
    }

    fun requestNotificationPermission(activity: Activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            ActivityCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {

            activity.requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), 1)
        }
    }

    fun showNotification(channelId: String, title: String, message: String) {
        val sharedPreferences = context.getSharedPreferences("settings", Context.MODE_PRIVATE)
        val pushNotificationsEnabled = sharedPreferences.getBoolean("push_notifications", true)

        // اگر اعلان‌ها غیرفعال شده‌اند، هیچ نوتیفیکیشنی ارسال نشود
        if (!pushNotificationsEnabled) {
            return
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            ActivityCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            return
        }

        val intent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        }
        val pendingIntent: PendingIntent = PendingIntent.getActivity(
            context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val largeIcon = BitmapFactory.decodeResource(context.resources, R.drawable.logo)

        val notificationId = (System.currentTimeMillis() % Int.MAX_VALUE).toInt()

        val builder = NotificationCompat.Builder(context, channelId)
            .setSmallIcon(R.drawable.notifsmall)
            .setLargeIcon(largeIcon)
            .setContentTitle(title)
            .setContentText(message)
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)

        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.notify(notificationId, builder.build())
    }

    fun showWelcomeNotification() {
        val builder = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.notifsmall) // اطمینان حاصل کنید که این آیکون در res/drawable وجود دارد
            .setContentTitle("Welcome")
            .setContentText("Welcome to ADL Wallet")
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)

        val notificationManager = NotificationManagerCompat.from(context)
        
        try {
            notificationManager.notify(NOTIFICATION_ID, builder.build())
        } catch (e: SecurityException) {
        }
    }
}





//fun onReceiveTransaction(amount: Double, currency: String) {
//    val notificationHelper = NotificationHelper(context)
//    notificationHelper.showNotification(
//        NotificationHelper.RECEIVE_CHANNEL_ID,
//        "دریافت وجه",
//        "شما $amount $currency دریافت کردید!"
//    )
//}
//
//
//
//fun onSendTransaction(amount: Double, currency: String) {
//    val notificationHelper = NotificationHelper(context)
//    notificationHelper.showNotification(
//        NotificationHelper.SEND_CHANNEL_ID,
//        "ارسال وجه",
//        "شما $amount $currency ارسال کردید!"
//    )
//}
//
//
//
//fun checkPriceAlert(currentPrice: Double, targetPrice: Double) {
//    if (currentPrice >= targetPrice) {
//        val notificationHelper = NotificationHelper(context)
//        notificationHelper.showNotification(
//            NotificationHelper.PRICE_ALERT_CHANNEL_ID,
//            "هشدار قیمت!",
//            "قیمت بیت‌کوین به $currentPrice دلار رسید!"
//        )
//    }
//}
